File: lc/valid-anagram.go
/*
can use one table of 26 chars -- map
s increments a given char, t decrements
if map is all 0 at end --> true. otherwise --> false
note: can return false imm if lengths are not the same
O(n) time, O(1) space

can sort both. then walk down. if s[i] != t[i] --> return false.
return true if never returned false.
O(n lg n) time, O(n) space

car rat
ss: acr
     i
ts: art
*/
func isAnagram(s string, t string) bool {
    if len(s) != len(t) {
        return false
    }
    sbs, tbs := []byte(s), []byte(t)
    slices.Sort(sbs)
    slices.Sort(tbs)
    for i := range sbs {
        if sbs[i] != tbs[i] {
            return false
        }
    }
    return true
}

File: lc/py/valid-anagram.py
# create two counters, and check if they are the same
# or use one counter, and ensure it is zero at the end: one increments, the other decrements
# either way: O(max(m,n)), O(1) <-- time and space
# a n a g r a m

# a b c 
# b a c 

# 0 0 0
# a b c
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        m,n = len(s), len(t)
        if m != n:
            return False
        c = [0 for _ in range(26)]
        for i in range(m):
            sch,tch = s[i],t[i]
            c[ord(sch) - ord('a')] += 1
            c[ord(tch) - ord('a')] -= 1
        return not any(bool(c[i]) for i in range(26))
